xc/device.ml

```diff
@@ -1430,6 +1431,116 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
 
 module Dm = struct
 
+  module QMP_Event = struct
+    let (pipe_r, pipe_w) = Unix.pipe ()
+    let channel_r = Unix.in_channel_of_descr pipe_r
+    let channel_w = Unix.out_channel_of_descr pipe_w
+
+    let epfd = Epoll.create1 [Epoll.EPOLL_CLOEXEC]
+
+    module Lookup = struct
+      let ftod, dtof = Hashtbl.create 16, Hashtbl.create 16
+
+      let remove domid =
+        if Hashtbl.mem dtof domid
+        then begin
+          try
+            let c = Hashtbl.find dtof domid in
+            Hashtbl.remove dtof domid;
+            Hashtbl.remove ftod c;
+            Epoll.ctl epfd Epoll.EPOLL_CTL_DEL (Qmp_protocol.to_fd c) [Epoll.EPOLLIN];
+            Qmp_protocol.close c
+          with
+          | e -> debug "Get exception trying remove qmp on domain-%d: %s" domid (Printexc.to_string e)
+        end
+
+      let add domid =
+        if not(Hashtbl.mem dtof domid)
+        then begin
+          try
+            let c = Qmp_protocol.connect (Printf.sprintf "/var/run/xen/qmp-event-%d" domid) in
+            Qmp_protocol.negotiate c;
+            Hashtbl.replace dtof domid c;
+            Hashtbl.replace ftod c domid;
+            Epoll.ctl epfd Epoll.EPOLL_CTL_ADD (Qmp_protocol.to_fd c) [Epoll.EPOLLIN];
+            debug "QMP domain-%d: negotiation complete" domid
+          with e ->
+            debug "QMP domain-%d: negotiation failed (%s): removing socket" domid (Printexc.to_string e);
+            remove domid
```

Comment:
I do not like this remove being called here, perhaps it's better to have 2 try: one internal in case Epoll.ctl fails, and another external in case Qmp_protocol.* fails.

Reply:
But `QMP_Event.remove` also has `try .. catch`, will it cause the thread fail if error happens?